### PR TITLE
=str #19072 fix FlattenMerge init race

### PR DESCRIPTION
**Pull request by rkuhn**

### CODE REVIEW

File: akka-http-core/src/main/java/akka/http/javadsl/model/MediaTypes.java [hunk 1]
```diff
@@ -190,7 +190,8 @@
     public static MediaType custom(String value, boolean binary, boolean compressible) {
         akka.http.scaladsl.model.MediaType.Compressibility comp = compressible ?
                 akka.http.scaladsl.model.MediaType.Compressible$.MODULE$ : akka.http.scaladsl.model.MediaType.NotCompressible$.MODULE$;
-        return akka.http.scaladsl.model.MediaType.custom(value, binary, comp , List.<String>empty());
+        return akka.http.scaladsl.model.MediaType.custom(value, binary, comp , 
+            akka.http.scaladsl.model.MediaType.custom$default$4());
```
Comment (rkuhn) [inline akka-http-core/src/main/java/akka/http/javadsl/model/MediaTypes.java]: `List.<String> empty()` is ambiguous in the eyes of the JDT … (/cc @sirthias)

Comment (sirthias) [inline akka-http-core/src/main/java/akka/http/javadsl/model/MediaTypes.java]: What's the JDT? And what kind of ambiguity does it see?

Comment (drewhk) [inline akka-http-core/src/main/java/akka/http/javadsl/model/MediaTypes.java]: It is the Eclipse Java compiler.

Comment (rkuhn) [inline akka-http-core/src/main/java/akka/http/javadsl/model/MediaTypes.java]: I don’t know what it “sees”, but it fails compilation.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 2]
```diff
@@ -145,29 +144,56 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
...
+  private var subF: Future[LocalSinkSubscription[T]] = _
+  private var sub: LocalSinkSubscription[T] = _
+
+  var elem: T = null.asInstanceOf[T]
+
+  def isActive: Boolean = sub ne null
+  def deactivate(): Unit = sub = null
+  def activate(f: Future[LocalSinkSubscription[T]]): Unit = {
+    subF = f
+    f.foreach(s ⇒ sub = s)(sameThreadExecutionContext)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: This looks extremely shady

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Good, I hate sunburns. Shall I add a comment that sheds some artificial light?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 3]
```diff
@@ -145,29 +144,56 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
...
  */
-private[fusing] final class LocalSource[T](var sub: LocalSinkSubscription[T] = null, var elem: T = null.asInstanceOf[T])
+private[fusing] final class LocalSource[T] {
+  private var subF: Future[LocalSinkSubscription[T]] = _
+  private var sub: LocalSinkSubscription[T] = _
+
+  var elem: T = null.asInstanceOf[T]
+
+  def isActive: Boolean = sub ne null
+  def deactivate(): Unit = sub = null
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Should this also null out the subF so it doesn't hang onto a dead ref?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 2 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Absolutely. What guarantees that `sub` is visible to the other threads when the completing thread has executed the callback?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 3 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Yes, it should, thanks!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 4]
```diff
@@ -145,29 +144,56 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
...
+  def activate(f: Future[LocalSinkSubscription[T]]): Unit = {
+    subF = f
+    f.foreach(s ⇒ sub = s)(sameThreadExecutionContext)
+  }
+
+  def pull(): Unit = {
+    if (sub eq null)
+      throw new IllegalStateException("not yet initialized, subF=" + subF.value)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Message could be made a bit less dodgy? `subF=...` is very weird.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 3 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: :+1:

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 5]
```diff
@@ -145,29 +144,56 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
...
+    sub.pull()
+  }
+
+  def cancel(): Unit = subF.foreach(_.cancel())(sameThreadExecutionContext)
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Note that `cancel` does not modify `subF` nor `sub`.
Also I believe there to be a potential race-condition between `cancel` and `activate` (as callbacks can run in any order).

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: No, LocalSource is only used from within the FlattenMerge logic, which is a sequential place.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: okay …

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 5 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: Now yes, but if those classes are ever reused by any other logic now all bets are off?

**Comment by rkuhn** [comment]
merged so that #19098 can work on top of it

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 5 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: This is why this is an implementation detail of FlattenMerge, which you are about to properly express by moving things into a tighter namespace (thanks!).

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 6]
```diff
@@ -145,29 +144,70 @@ private[fusing] final class LocalSinkSubscription[T](sub: ActorPublisherMessage
...
+  def activate(f: Future[LocalSinkSubscription[T]]): Unit = {
+    subF = f
+    /*
+     * The subscription is communicated to the FlattenMerge stage by way of completing
+     * the future. Encoding it like this means that the `sub` field will be written
+     * either by us (if the future has already been completed) or by the LocalSink (when
+     * it eventually completes the future in its `preStart`). The important part is that
+     * either way the `sub` field is populated before we get the first `OnNext` message
+     * and the value is safely published in either case as well (since AsyncCallback is
+     * based on an Actor message send).
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: much needed comment, thanks for adding it :+1:

**Pull request merged**